File: apps/kinh-dich-service/src/domain/repositories.ts
/**
 * Kinh Dich Service — Repository Ports (domain interfaces)
 *
 * Pure interfaces. Implementations live in infrastructure/.
 */

import type { KinhDichStoredRow, MarkovData } from './models.js';

// ─── History row returned from kinhdich_readings ───────────────────────────────
export interface KinhDichHistoryRow {
  timestamp: string;
  hexagram_number: number;
  trading_signal: string | null;
  confidence: number | null;
}

// ─── Transition row returned from hexagram_transitions or similar table ────────
export interface KinhDichTransitionRow {
  to_hexagram: number;
  count: number;
  probability: number;
}

// ─── Price row for backtest ────────────────────────────────────────────────────
export interface KinhDichPriceRow {
  price: number;
  timestamp: string;
}

/** Port for reading Kinh Dich stored readings from a data store. */
export interface KinhDichRepositoryPort {
  /** Get the most recent stored reading for a given stock code. */
  getLatestReading(stockCode: string): KinhDichStoredRow | null;

  /** Get Markov transition data for a hexagram number. */
  getMarkovData(hexagramNumber: number): MarkovData | null;

  /** Get all readings for a stock within the past N days (for history + backtest). */
  getReadingsHistory(stockCode: string, days: number): KinhDichHistoryRow[];

  /** Get top-N transitions from a given hexagram for a stock (or all stocks). */
  getTopTransitions(fromHexagram: number, stockCode: string, topN: number): KinhDichTransitionRow[];

  /** Get price rows for a stock within the past N days (for backtest). */
  getPriceHistory(stockCode: string, days: number): KinhDichPriceRow[];
}

/** Port for computing hexagram scores from price history. */
export interface PriceScorePort {
  /** Compute 6 normalised scores from price history for a stock. Returns null if insufficient data. */
  computeScores(stockCode: string, days: number): number[] | null;
}

File: apps/kinh-dich-service/pkg/infrastructure/repositories.go
// Package infrastructure contains the infrastructure layer implementations.
package infrastructure

import (
	"os"

	"github.com/vn-market-intelligence/kinh-dich-service/pkg/domain"
)

// SQLiteReadingRepository implements KinhDichRepositoryPort using modernc.org/sqlite.
type SQLiteReadingRepository struct {
	dbPath string
}

// NewSQLiteReadingRepository creates a new SQLiteReadingRepository.
// Reads DB_PATH from environment variable (no-arg constructor per OQ-2).
func NewSQLiteReadingRepository() *SQLiteReadingRepository {
	dbPath := os.Getenv("DB_PATH")
	if dbPath == "" {
		dbPath = "./market.db" // default for local dev
	}
	return &SQLiteReadingRepository{
		dbPath: dbPath,
	}
}

// GetLatestReading returns the most recent stored reading for a given stock code.
// Stub implementation - real DB access in Phase 2.
func (r *SQLiteReadingRepository) GetLatestReading(stockCode string) *domain.KinhDichStoredRow {
	// Stub: return nil (no stored reading)
	return nil
}

// GetMarkovData returns Markov transition data for a hexagram number.
// Stub implementation - real DB access in Phase 2.
func (r *SQLiteReadingRepository) GetMarkovData(hexagramNumber int) *domain.MarkovData {
	// Stub: return nil (no Markov data)
	return nil
}

// GetReadingsHistory returns all readings for a stock within the past N days.
// Stub implementation - real DB access in Phase 2.
func (r *SQLiteReadingRepository) GetReadingsHistory(stockCode string, days int) []domain.KinhDichHistoryRow {
	// Stub: return empty slice
	return nil
}

// GetTopTransitions returns top-N transitions from a given hexagram for a stock.
// Stub implementation - real DB access in Phase 2.
func (r *SQLiteReadingRepository) GetTopTransitions(fromHexagram int, stockCode string, topN int) []domain.KinhDichTransitionRow {
	// Stub: return empty slice
	return nil
}

// GetPriceHistory returns price rows for a stock within the past N days.
// Stub implementation - real DB access in Phase 2.
// Implements PriceHistorySource interface (returns nil, nil for stub).
func (r *SQLiteReadingRepository) GetPriceHistory(stockCode string, days int) ([]domain.KinhDichPriceRow, error) {
	// Stub: return empty slice (not an error, just no local data)
	return nil, nil
}
